### PR TITLE
Decouple chat presentation and tracking

### PR DESCRIPTION
**Pull request by unknownbrackets**
A few benefits to this model:
 * Possibly multiple parts of the UI could listen to chat messages - for example, an audio cue or app badge, etc.
 * There seemed to be threading concerns depending on when exactly people chat before, this cleans that up.
 * UI visibility no longer part of the equation for the tracking side - less likely to have bugs like #14723 in the future.

Alternative to #14854 (sorry), but good catch on the cause, thanks.  Closes #14854.

-[Unknown]

### CODE REVIEW

**Comment by hrydgard** [comment]
I'm actually not 100% sure the threading concern is real, I don't understand the AdHoc code well enough :) But I think it's highly unlikely that the mutex locking here will cause any performance problem, so let's have it.

**Comment by unknownbrackets** [comment]
As far as I can tell, `friendFinder()` runs on a separate thread and modifies `chatLog`,  It's started in sceNetAdhoc.cpp by `friendFinderThread = std::thread(friendFinder);`.  So definitely seems like a potential concern, as before it could push_back twice concurrently, erase, or copy out of it during the push_back.

-[Unknown]

File: UI/EmuScreen.cpp [hunk 1]
```diff
@@ -987,6 +987,18 @@ void EmuScreen::update() {
 	resumeButton_->SetVisibility(coreState == CoreState::CORE_RUNTIME_ERROR && Memory::MemFault_MayBeResumable() ? V_VISIBLE : V_GONE);
 	resetButton_->SetVisibility(coreState == CoreState::CORE_RUNTIME_ERROR ? V_VISIBLE : V_GONE);
 
+	if (chatButton_ && chatMenu_) {
+		if (chatMenu_->GetVisibility() != V_GONE) {
+			chatMessages_ = GetChatMessageCount();
+			newChatMessages_ = 0;
+		} else {
+			int diff = GetChatMessageCount() - chatMessages_;
+			chatMessages_ += diff;
```
Comment (iota97) [inline UI/EmuScreen.cpp:996]: `chatMessages_` should not be updated or it will display 0 the next frame.

File: Core/HLE/proAdhoc.cpp [hunk 2]
```diff
@@ -1290,42 +1289,47 @@ void sendChat(std::string chatString) {
...
 			//maximum char allowed is 64 character for compability with original server (pro.coldbird.net)
-			message = chatString.substr(0, 60); // 64 return chat variable corrupted is it out of memory?
+			std::string message = chatString.substr(0, 60); // 64 return chat variable corrupted is it out of memory?
 			strcpy(chat.message, message.c_str());
 			//Send Chat Messages
 			if (IsSocketReady(metasocket, false, true) > 0) {
 				int chatResult = send(metasocket, (const char*)&chat, sizeof(chat), MSG_NOSIGNAL);
 				NOTICE_LOG(SCENET, "Send Chat %s to Adhoc Server", chat.message);
-				name = g_Config.sNickName.c_str();
+				std::string name = g_Config.sNickName.c_str();
```
Comment (iota97) [inline Core/HLE/proAdhoc.cpp:1302]: nit: c_str() is not needed

File: UI/EmuScreen.cpp (hunk 1 above)
Comment (unknownbrackets) [reply UI/EmuScreen.cpp:996]: Not sure what I was thinking, I added that afterward but I really can't remember why and I somehow thought it was needed as I was typing it.  Must be my insanity finally taking over.

-[Unknown]

File: Core/HLE/proAdhoc.cpp (hunk 2 above)
Comment (unknownbrackets) [reply Core/HLE/proAdhoc.cpp:1302]: Ah, yeah, I was just trying to move it from a global so it wouldn't be a threading hazard.

-[Unknown]

**Pull request merged**